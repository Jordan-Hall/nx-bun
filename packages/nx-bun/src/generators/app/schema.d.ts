import type { ProjectNameAndRootFormat } from '@nx/devkit/src/generators/project-name-and-root-utils';
export interface AppGeneratorSchema {
  name: string;
  applicationType?: AppTypes;
  directory?: string;
  projectNameAndRootFormat?: ProjectNameAndRootFormat;
  tags?: string;
  rootProject?: boolean;
}
export type AppTypes = 'api' | 'none';