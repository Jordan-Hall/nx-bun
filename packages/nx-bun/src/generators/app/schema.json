{
  "$schema": "http://json-schema.org/schema",
  "$id": "App",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "applicationType": {
      "description": "Generate the node application using a framework",
      "type": "string",
      "enum": ["api", "none"],
      "default": "api",
      "x-prompt": "What of app/framework do you want to use?",
      "x-priority": "important"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "x-priority": "important"
    },
    "projectNameAndRootFormat": {
      "description": "Whether to generate the project name and root directory as provided (`as-provided`) or generate them composing their values and taking the configured layout into account (`derived`).",
      "type": "string",
      "enum": ["as-provided", "derived"]
    },
    "rootProject": {
      "description": "Create node application at the root of the workspace",
      "type": "boolean",
      "default": false,
      "hidden": true,
      "x-priority": "internal"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the library (used for linting)."
    }
  },
  "required": ["name"]
}
