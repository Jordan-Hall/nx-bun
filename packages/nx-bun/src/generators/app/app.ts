import {
  addProjectConfiguration,
  formatFiles,
  generateFiles,
  readJson,
  names,
  Tree,
  TargetConfiguration,
  joinPathFragments,
  offsetFromRoot,
} from '@nx/devkit';

import * as path from 'path';
import {
  determineProjectNameAndRootOptions,
  type ProjectNameAndRootOptions,
} from '@nx/devkit/src/generators/project-name-and-root-utils';
import { AppGeneratorSchema, AppTypes } from './schema';
import { BundleExecutorSchema } from '../../executors/build/schema';
import { TestExecutorSchema } from '../../executors/test/schema';
import { getRootTsConfigPathInTree } from '../../utils/ts-config';
import { RunExecutorSchema } from '../../executors/run/schema';
import initGenerator from '../init/init';


export interface NormalizedSchema extends AppGeneratorSchema {
  name: string;
  projectNames: ProjectNameAndRootOptions['names'];
  fileName: string;
  projectRoot: ProjectNameAndRootOptions['projectRoot'];
  projectName: ProjectNameAndRootOptions['projectName'];
  parsedTags: string[];
  importPath?: ProjectNameAndRootOptions['importPath']
  propertyName: string
}

type AppStructure = {
  genFiles: () => void;
  packageInstall: () => void;
};
type ApplicationGeneration =  Partial<Record<AppTypes, AppStructure>>;
export async function appGenerator(tree: Tree, options: AppGeneratorSchema) {
  const opts = await normalizeOptions(tree, options);

  await initGenerator(tree, {bunNXRuntime: false, forceBunInstall: false})

  const entryPoints =  [joinPathFragments(opts.projectRoot, 'src', 'main.ts')]

  const build: TargetConfiguration<BundleExecutorSchema> = {
    executor: '@nx-bun/nx:build',
    outputs: ["{options.outputPath}"],
    options: {
      entrypoints: entryPoints,
      outputPath: joinPathFragments(
        'dist',
        opts.projectRoot ? opts.name : opts.projectRoot,
      ),
      smol: false,
      bun: true
    }
  }

  const serve: TargetConfiguration<RunExecutorSchema> = {
    executor: '@nx-bun/nx:run',
    defaultConfiguration: "development",
    options: {
      main: joinPathFragments(opts.projectRoot, 'src', 'main.ts'),
      watch: true,
      hot: true,
      bun: true,
      smol: false
    }
  }

  const test: TargetConfiguration<TestExecutorSchema> = {
    executor: '@nx-bun/nx:test',
    options: {
      smol: false,
      bail: true,
      bun: true
    }
  }


  addProjectConfiguration(tree, options.name, {
    root: opts.projectRoot,
    projectType: 'application',
    sourceRoot: `${opts.projectRoot}/src`,
    targets: {
      build,
      serve,
      test
    },
  });

  createFiles(tree, opts)
  

  await formatFiles(tree);
}

async function normalizeOptions(
  tree: Tree,
  options: AppGeneratorSchema
): Promise<NormalizedSchema> {

  const {
    projectName,
    names: projectNames,
    projectRoot,
    importPath,
  } = await determineProjectNameAndRootOptions(tree, {
    name: options.name,
    projectType: 'application',
    directory: options.directory,
  });


  const fileName = getCaseAwareFileName({
    fileName: projectNames.projectFileName,
    pascalCaseFiles: false
  });

  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    fileName,
    name: projectName,
    projectNames,
    projectName,
    propertyName: names(fileName).propertyName,
    projectRoot,
    parsedTags,
    importPath,
  };
}

function getCaseAwareFileName(options: {
  pascalCaseFiles: boolean;
  fileName: string;
}) {
  const normalized = names(options.fileName);

  return options.pascalCaseFiles ? normalized.className : normalized.fileName;
}


function createFiles(tree: Tree, opts: NormalizedSchema) {

  const templateOptions = {
    ...opts,
    template: '',
    cliCommand: 'nx',
    offsetFromRoot: offsetFromRoot(opts.projectRoot),
    baseTsConfig: getRootTsConfigPathInTree(tree)
  };

  generateFiles(tree, path.join(__dirname, 'files/common'), `${opts.projectRoot}`, templateOptions);

  const applicationType: ApplicationGeneration = {
    'api': {
      genFiles: () => {
        tree.delete(joinPathFragments(opts.projectRoot, 'src', 'main.ts'));
        generateFiles(tree, path.join(__dirname, 'files/api'), `${opts.projectRoot}`, templateOptions);
      },
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      packageInstall: () => {}
    },
  }

  const genAppDetails = applicationType[opts.applicationType];
  genAppDetails?.genFiles();
  genAppDetails?.packageInstall();
  
}

export default appGenerator;

