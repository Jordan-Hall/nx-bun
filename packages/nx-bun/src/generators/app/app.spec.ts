import { readJson, readProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { appGenerator } from './app'
import { getRootTsConfigPathInTree } from '@nx/js';

describe('Bun App Generator', () => {
	let tree: Tree;
	describe('General creation', () => {
		beforeEach(() => {
			tree = createTreeWithEmptyWorkspace();
		});

		it('should make a bun app', async () => {
      await appGenerator(tree, {
        name: 'my-app',
      });
			expect(readJson(tree, getRootTsConfigPathInTree(tree))?.compilerOptions?.types).toContain('bun-types')
      expect(readProjectConfiguration(tree, 'my-app')).toMatchSnapshot();
      expect(tree.exists('src/index.ts')).toBeTruthy();
    });
	});

	describe('integrated', () => {
    beforeEach(() => {
      tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    });

    it('should make an integrated bun app', async () => {
      await appGenerator(tree, {
        name: 'my-app',
      });
			
      expect(readJson(tree, getRootTsConfigPathInTree(tree))?.compilerOptions?.types).toContain('bun-types')
      expect(readProjectConfiguration(tree, 'my-app')).toMatchSnapshot();
      expect(tree.exists('apps/my-app/src/index.ts')).toBeTruthy();
    });
	})
})
