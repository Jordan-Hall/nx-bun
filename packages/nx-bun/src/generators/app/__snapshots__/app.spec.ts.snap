// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bun App Generator General creation should generate a new application project with the given name and default options 1`] = `
{
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "name": "my-app",
  "projectType": "application",
  "root": "my-app",
  "sourceRoot": "my-app/src",
  "targets": {
    "build": {
      "executor": "@nx-bun/nx:build",
      "options": {
        "bun": true,
        "entrypoints": [
          "my-app/src/main.ts",
        ],
        "outputPath": "dist/my-app",
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "serve": {
      "defaultConfiguration": "development",
      "executor": "@nx-bun/nx:run",
      "options": {
        "buildTarget": "my-app:build",
        "bun": true,
        "hot": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx-bun/nx:test",
      "options": {
        "bail": true,
        "bun": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.json",
      },
    },
  },
}
`;

exports[`Bun App Generator General creation should generate a new application project with the given name and specified options and application type 1`] = `
{
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "name": "my-app",
  "projectType": "application",
  "root": "my-app",
  "sourceRoot": "my-app/src",
  "targets": {
    "build": {
      "executor": "@nx-bun/nx:build",
      "options": {
        "bun": true,
        "entrypoints": [
          "my-app/src/main.ts",
        ],
        "outputPath": "dist/my-app",
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "serve": {
      "defaultConfiguration": "development",
      "executor": "@nx-bun/nx:run",
      "options": {
        "buildTarget": "my-app:build",
        "bun": true,
        "hot": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx-bun/nx:test",
      "options": {
        "bail": true,
        "bun": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.json",
      },
    },
  },
}
`;

exports[`Bun App Generator General creation should generate a new application project with the given name and specified options and directory 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "my-app",
  "projectType": "application",
  "root": "apps/my-app",
  "sourceRoot": "apps/my-app/src",
  "targets": {
    "build": {
      "executor": "@nx-bun/nx:build",
      "options": {
        "bun": true,
        "entrypoints": [
          "apps/my-app/src/main.ts",
        ],
        "outputPath": "dist/my-app",
        "smol": false,
        "tsConfig": "apps/my-app/tsconfig.app.json",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "serve": {
      "defaultConfiguration": "development",
      "executor": "@nx-bun/nx:run",
      "options": {
        "buildTarget": "my-app:build",
        "bun": true,
        "hot": true,
        "smol": false,
        "tsConfig": "apps/my-app/tsconfig.app.json",
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx-bun/nx:test",
      "options": {
        "bail": true,
        "bun": true,
        "smol": false,
        "tsConfig": "apps/my-app/tsconfig.json",
      },
    },
  },
}
`;

exports[`Bun App Generator General creation should generate a new application project with the given name and specified options and tags 1`] = `
{
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "name": "my-app",
  "projectType": "application",
  "root": "my-app",
  "sourceRoot": "my-app/src",
  "targets": {
    "build": {
      "executor": "@nx-bun/nx:build",
      "options": {
        "bun": true,
        "entrypoints": [
          "my-app/src/main.ts",
        ],
        "outputPath": "dist/my-app",
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "serve": {
      "defaultConfiguration": "development",
      "executor": "@nx-bun/nx:run",
      "options": {
        "buildTarget": "my-app:build",
        "bun": true,
        "hot": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx-bun/nx:test",
      "options": {
        "bail": true,
        "bun": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.json",
      },
    },
  },
}
`;

exports[`Bun App Generator General creation should make a bun app 1`] = `
{
  "$schema": "../node_modules/nx/schemas/project-schema.json",
  "name": "my-app",
  "projectType": "application",
  "root": "my-app",
  "sourceRoot": "my-app/src",
  "targets": {
    "build": {
      "executor": "@nx-bun/nx:build",
      "options": {
        "bun": true,
        "entrypoints": [
          "my-app/src/main.ts",
        ],
        "outputPath": "dist/my-app",
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "serve": {
      "defaultConfiguration": "development",
      "executor": "@nx-bun/nx:run",
      "options": {
        "buildTarget": "my-app:build",
        "bun": true,
        "hot": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.app.json",
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx-bun/nx:test",
      "options": {
        "bail": true,
        "bun": true,
        "smol": false,
        "tsConfig": "my-app/tsconfig.json",
      },
    },
  },
}
`;

exports[`Bun App Generator integrated should make an integrated bun app 1`] = `
{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "name": "my-app",
  "projectType": "application",
  "root": "apps/my-app",
  "sourceRoot": "apps/my-app/src",
  "targets": {
    "build": {
      "executor": "@nx-bun/nx:build",
      "options": {
        "bun": true,
        "entrypoints": [
          "apps/my-app/src/main.ts",
        ],
        "outputPath": "dist/my-app",
        "smol": false,
        "tsConfig": "apps/my-app/tsconfig.app.json",
      },
      "outputs": [
        "{options.outputPath}",
      ],
    },
    "serve": {
      "defaultConfiguration": "development",
      "executor": "@nx-bun/nx:run",
      "options": {
        "buildTarget": "my-app:build",
        "bun": true,
        "hot": true,
        "smol": false,
        "tsConfig": "apps/my-app/tsconfig.app.json",
        "watch": true,
      },
    },
    "test": {
      "executor": "@nx-bun/nx:test",
      "options": {
        "bail": true,
        "bun": true,
        "smol": false,
        "tsConfig": "apps/my-app/tsconfig.json",
      },
    },
  },
}
`;
